data/tutorials/language/5rt_01_garbage-collector.md

```diff
@@ -0,0 +1,677 @@
+When you execute your program, OCaml manages the lifecycle of these variables
+by regularly scanning allocated values and freeing them when they're no
+longer needed. This in turn means that your applications don't need to
+manually implement memory management, and it greatly reduces the likelihood
+of memory leaks creeping into your code.
+
+The OCaml runtime is a C library that provides routines that can be called
+from running OCaml programs. The runtime manages a *heap*, which is a
+collection of memory regions that it obtains from the operating system. The
+runtime uses this memory to hold *heap blocks* that it fills up with OCaml
+values in response to allocation requests by the OCaml program.
+
+## Mark and Sweep Garbage Collection
+
+When there isn't enough memory available to satisfy an allocation request
+from the pool of allocated heap blocks, the runtime system invokes the
+garbage collector (GC). An OCaml program can't explicitly free a value when
+it is done with it. Instead, the GC regularly determines which values are
+*live* and which values are *dead*, i.e., no longer in use. Dead values are
+collected and their memory made available for reuse by the application.
+
+The GC doesn't keep constant track of values as they are allocated and used.
+Instead, it regularly scans them by starting from a set of *root* values that
+the application always has access to (such as the stack). The GC maintains a
+directed graph in which heap blocks are nodes, and there is an edge from heap
+block `b1` to heap block `b2` if some field of `b1` is a pointer to `b2`.
+
+All blocks reachable from the roots by following edges in the graph must be
+retained, and unreachable blocks can be reused by the application. The
+algorithm used by OCaml to perform this heap traversal is commonly known as
+*mark and sweep* garbage collection, and we'll explain it further now.
+
+## Generational Garbage Collection
+
+The usual OCaml programming style involves allocating many small
+values that are used for a short period of time and then never
+accessed again. OCaml takes advantage of this fact to improve
+performance by using a *generational* GC.
+
+A generational GC maintains separate memory regions to hold blocks based on
+how long the blocks have been live. OCaml's heap is split into two such
+regions:
+
+- A small, fixed-size *minor heap* where most blocks are initially allocated
+
+- A larger, variable-size *major heap* for blocks that have been live longer
+
+A typical functional programming style means that young blocks tend to
+die young and old blocks tend to stay around for longer than young
+ones. This is often referred to as the *generational
+hypothesis*.
+
+OCaml uses different memory layouts and garbage-collection algorithms for the
+major and minor heaps to account for this generational difference. We'll
+explain how they differ in more detail next.
+
+<div class="note">
```

Comment:
Do these notes render properly?

Reply:
yes, I added styles to render them removed from the main text flow